src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+            foreach (ref readonly T item in items)
+            {
+                multiSet.Add(item);
+            }
+
+            return GetOrderedIndicesToRemoveFor(multiSet);
+        }
+
+        private ICollection<int>? GetOrderedIndicesToRemoveFor(MultiSet multiSet)
+        {
+            List<int>? indicesToRemove = null;
+            for (int i = 0; i < array!.Length; i++)
+            {
+                if (multiSet.TryRemove(array[i]))
+                {
+                    (indicesToRemove ??= new()).Add(i);
+                }
+            }
+
+            return indicesToRemove;
+        }
+
+        private SortedSet<int> FindOrderedIndicesToRemove(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
+        {
+            var indicesToRemove = new SortedSet<int>();
+            foreach (T item in items)
+            {
+                int index = -1;
+                do
+                {
+                    index = IndexOf(item, index + 1, equalityComparer);
+                } while (index >= 0 && !indicesToRemove.Add(index) && index < Length - 1);
+            }
+
+            return indicesToRemove;
+        }
+
+        private class MultiSet
+        {
+            private readonly IEqualityComparer<T> _equalityComparer;
+            private int[] _bucketHeaders; // initialized as zeroed array so its valid item value is 1-based index of _entries
+            private Entry[] _entries;
+            private int _startOffset;
+
+            public MultiSet(int capacity, IEqualityComparer<T>? equalityComparer)
+            {
+                _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
+
+                _bucketHeaders = new int[capacity];
+                _entries = new Entry[capacity];
+            }
+
+            public MultiSet(IEqualityComparer<T>? equalityComparer)
+            {
+                _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
+
+                _bucketHeaders = new int[1];
+                _entries = new Entry[1];
+            }
+
+            public void Add(T item)
+            {
+                ref int bucketHeader = ref GetBucketHeaderIndexRef(item);
+                int entryIndex = bucketHeader - 1;
+                while (entryIndex != -1)
+                {
+                    ref Entry entry = ref _entries[entryIndex];
+                    if (_equalityComparer.Equals(entry.Key, item))
+                    {
+                        entry.Count++;
+                        return;
+                    }
+
+                    entryIndex = entry.Next;
+                }
+
+                if (_startOffset == _bucketHeaders.Length)
+                {
+                    EnsureCapacity();
+                    bucketHeader = ref GetBucketHeaderIndexRef(item);
+                }
+
+                _entries[_startOffset].SetInfo(item, bucketHeader - 1, 1);
+                bucketHeader = ++_startOffset; // 1-based bucketHeader
```

Comment:
Why does it need to be assigned to a local?

Reply:
`bucketHeader` is a `ref` variable to actual bucket header in `_bucketHeaders` for `item`. Here it needs to be updated to a new offset.